File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
from collections import deque

n, m,v = map(int, input().split())

#그래프를 리스트 자료형으로 표현
graph=[]
for i in range(1001):
  graph.append([])

#각 노드가 연결된 정보를 표현
for i in range(m):
  a, b =map(int,input().split())
  graph[a].append(b)
  graph[b].append(a)

#각 노드가 방문된 정보를 리스트 자료형으로 표현
visited=[False]*1001
visited_bfs=[False]*1001

#dfs 메소드 정의
def dfs(graph, v, visited):
  #현재 노드를 방문 처리
  visited[v]=True
  print(v, end=' ')
  #번호가 작은 노드부터 먼저 방문하기 위해 숫자 정렬
  graph[v].sort()
  #현재 노드와 연결된 노드 중 방문한 적이 없는 노드를 재귀적으로 방문
  for i in graph[v]:
    if not visited[i]:
      dfs(graph,i, visited)

#bfs 메소드 정의
def bfs(graph, v, visited):
  #큐 구현을 위해 deque 라이브러리 이용
  queue=deque([v])
  #현재 노드를 방문 처리
  visited[v]=True
  #큐가 빌 때까지 반복한다
  while queue:
    #큐에서 하나의 원소를 뽑아 출력한다
    v=queue.popleft()
    print(v, end=' ')
    #번호가 작은 노드부터 먼저 방문하기 위해 숫자 정렬
    graph[v].sort()
    #현재 노드에 연결된 노드 중 방문한 적이 없는 노드를 큐에 삽입&방문 처리
    for i in graph[v]:
      if not visited[i]:
        queue.append(i)
        visited[i]=True
  

dfs(graph, v, visited)
print()
bfs(graph, v, visited_bfs)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.*;
import java.util.*;

public class Main {
    static boolean visited[];
    static ArrayList<Integer>[] arr;
  public static void main(String args[]) throws IOException {
      BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
      StringTokenizer st = new StringTokenizer(br.readLine());
      
      int N = Integer.parseInt(st.nextToken());
      int M = Integer.parseInt(st.nextToken());
      int V = Integer.parseInt(st.nextToken());
      
      arr = new ArrayList[N+1];
      for(int i=1; i<N+1; i++){
          arr[i]=new ArrayList<Integer>();
      }
      for(int i=0; i<M; i++){
          st = new StringTokenizer(br.readLine());
          int a = Integer.parseInt(st.nextToken());
          int b = Integer.parseInt(st.nextToken());
          arr[a].add(b);
          arr[b].add(a);
      }
      //번호가 작은 것을 먼저 방문하기 위해 정렬
      for(int i=1;i<N+1;i++){
          Collections.sort(arr[i]);
      }
      visited = new boolean[N+1];
      dfs(V);
      System.out.println();
      visited = new boolean[N+1];
      bfs(V);
      
      br.close();
  }
  
  static void dfs(int node){
      System.out.print(node+" ");
      visited[node] = true;
      for(int i: arr[node]){
          if(!visited[i]){
              dfs(i);
          }
      }
  }
  
  static void bfs(int node){
      Queue<Integer> queue = new ArrayDeque<>();
      queue.add(node);
      visited[node]=true;
      
      while(!queue.isEmpty()){
          int v = queue.poll();
          System.out.print(v+" ");
          for(int i : arr[v]){
              if(!visited[i]){
                  visited[i]=true;
                  queue.add(i);
              }
          }
      }
  }
}